From f809dbaf61e19d780c9421bd74f7b04aa29b4d8f Mon Sep 17 00:00:00 2001
From: Stefan Saraev <stefan@saraev.ca>
Date: Thu, 19 Dec 2013 11:21:13 +0200
Subject: [PATCH 1/5] move /etc/systemd/system to /storage/.config/system.d

---
 Makefile.am | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 3b7cc1e..b92465b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -185,7 +185,7 @@ CLEANFILES = $(BUILT_SOURCES) \
 AM_CPPFLAGS = \
 	-include $(top_builddir)/config.h \
 	-DPKGSYSCONFDIR=\"$(pkgsysconfdir)\" \
-	-DSYSTEM_CONFIG_UNIT_PATH=\"$(pkgsysconfdir)/system\" \
+	-DSYSTEM_CONFIG_UNIT_PATH=\"/storage/.config/system.d\" \
 	-DSYSTEM_DATA_UNIT_PATH=\"$(systemunitdir)\" \
 	-DSYSTEM_SYSVINIT_PATH=\"$(SYSTEM_SYSVINIT_PATH)\" \
 	-DSYSTEM_SYSVRCND_PATH=\"$(SYSTEM_SYSVRCND_PATH)\" \
@@ -5916,7 +5916,7 @@ substitutions = \
        '|SYSTEMCTL=$(rootbindir)/systemctl|' \
        '|SYSTEMD_NOTIFY=$(rootbindir)/systemd-notify|' \
        '|pkgsysconfdir=$(pkgsysconfdir)|' \
-       '|SYSTEM_CONFIG_UNIT_PATH=$(pkgsysconfdir)/system|' \
+       '|SYSTEM_CONFIG_UNIT_PATH=/storage/.config/system.d|' \
        '|USER_CONFIG_UNIT_PATH=$(pkgsysconfdir)/user|' \
        '|pkgdatadir=$(pkgdatadir)|' \
        '|systemunitdir=$(systemunitdir)|' \
-- 
2.1.4


From 8223882f5d0770e8b6c22e4518de9faff54c8e85 Mon Sep 17 00:00:00 2001
From: Stefan Saraev <stefan@saraev.ca>
Date: Fri, 7 Aug 2015 13:25:23 +0300
Subject: [PATCH 2/5] move hwdb.bin to /run

---
 src/hwdb/hwdb.c                  | 2 +-
 src/libsystemd/sd-hwdb/sd-hwdb.c | 2 +-
 src/udev/udevadm-hwdb.c          | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/hwdb/hwdb.c b/src/hwdb/hwdb.c
index 1160dac..bb09628 100644
--- a/src/hwdb/hwdb.c
+++ b/src/hwdb/hwdb.c
@@ -41,7 +41,7 @@
  * Uses a Patricia/radix trie to index all matches for efficient lookup.
  */
 
-static const char *arg_hwdb_bin_dir = "/etc/udev";
+static const char *arg_hwdb_bin_dir = "/run";
 static const char *arg_root = "";
 
 static const char * const conf_file_dirs[] = {
diff --git a/src/libsystemd/sd-hwdb/sd-hwdb.c b/src/libsystemd/sd-hwdb/sd-hwdb.c
index 062fa97..90ef3b6 100644
--- a/src/libsystemd/sd-hwdb/sd-hwdb.c
+++ b/src/libsystemd/sd-hwdb/sd-hwdb.c
@@ -271,7 +271,7 @@ static int trie_search_f(sd_hwdb *hwdb, const char *search) {
 
 static const char hwdb_bin_paths[] =
         "/etc/systemd/hwdb/hwdb.bin\0"
-        "/etc/udev/hwdb.bin\0"
+        "/run/hwdb.bin\0"
         "/usr/lib/systemd/hwdb/hwdb.bin\0"
 #ifdef HAVE_SPLIT_USR
         "/lib/systemd/hwdb/hwdb.bin\0"
diff --git a/src/udev/udevadm-hwdb.c b/src/udev/udevadm-hwdb.c
index 948ad0f..a14cb98 100644
--- a/src/udev/udevadm-hwdb.c
+++ b/src/udev/udevadm-hwdb.c
@@ -564,7 +564,7 @@ static int adm_hwdb(struct udev *udev, int argc, char *argv[]) {
         };
         const char *test = NULL;
         const char *root = "";
-        const char *hwdb_bin_dir = "/etc/udev";
+        const char *hwdb_bin_dir = "/run";
         bool update = false;
         struct trie *trie = NULL;
         int err, c;
-- 
2.1.4


From f03cb82529543e0738116d53a8eb7b0e284a4147 Mon Sep 17 00:00:00 2001
From: Stefan Saraev <stefan@saraev.ca>
Date: Sun, 8 Nov 2015 11:54:39 +0200
Subject: [PATCH 3/5] always run machine-id-commit

---
 units/systemd-machine-id-commit.service.in | 2 --
 1 file changed, 2 deletions(-)

diff --git a/units/systemd-machine-id-commit.service.in b/units/systemd-machine-id-commit.service.in
index 1f3f5da..90b346c 100644
--- a/units/systemd-machine-id-commit.service.in
+++ b/units/systemd-machine-id-commit.service.in
@@ -12,8 +12,6 @@ DefaultDependencies=no
 Conflicts=shutdown.target
 Before=sysinit.target shutdown.target
 After=local-fs.target
-ConditionPathIsReadWrite=/etc
-ConditionPathIsMountPoint=/etc/machine-id
 
 [Service]
 Type=oneshot
-- 
2.1.4


From 880d553c622b84ad7ed00e536ed23619ced796ca Mon Sep 17 00:00:00 2001
From: Stefan Saraev <stefan@saraev.ca>
Date: Sun, 29 Nov 2015 11:41:32 +0200
Subject: [PATCH 4/5] always run systemd-hwdb-update. run before udev

---
 units/systemd-hwdb-update.service.in | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/units/systemd-hwdb-update.service.in b/units/systemd-hwdb-update.service.in
index 7135cff..ae9e5fa 100644
--- a/units/systemd-hwdb-update.service.in
+++ b/units/systemd-hwdb-update.service.in
@@ -11,11 +11,7 @@ Documentation=man:hwdb(7) man:systemd-hwdb(8)
 DefaultDependencies=no
 Conflicts=shutdown.target
 After=systemd-remount-fs.service
-Before=sysinit.target shutdown.target systemd-update-done.service
-ConditionNeedsUpdate=/etc
-ConditionPathExists=|!@udevlibexecdir@/hwdb.bin
-ConditionPathExists=|/etc/udev/hwdb.bin
-ConditionDirectoryNotEmpty=|/etc/udev/hwdb.d/
+Before=systemd-udevd.service
 
 [Service]
 Type=oneshot
-- 
2.1.4


From 404320a7c3278b9db83906c57142110544ad6414 Mon Sep 17 00:00:00 2001
From: Stefan Saraev <stefan@saraev.ca>
Date: Wed, 18 Nov 2015 20:15:43 +0200
Subject: [PATCH 5/5] meh btrfs

---
 src/basic/btrfs-util.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/basic/btrfs-util.c b/src/basic/btrfs-util.c
index 359d85f..44a3723 100644
--- a/src/basic/btrfs-util.c
+++ b/src/basic/btrfs-util.c
@@ -1119,9 +1119,6 @@ int btrfs_quota_scan_start(int fd) {
 int btrfs_quota_scan_wait(int fd) {
         assert(fd >= 0);
 
-        if (ioctl(fd, BTRFS_IOC_QUOTA_RESCAN_WAIT) < 0)
-                return -errno;
-
         return 0;
 }
 
-- 
2.1.4

