diff --git a/sshconnect2.c b/sshconnect2.c
index f79c96b..6c8cdd3 100644
--- a/sshconnect2.c
+++ b/sshconnect2.c
@@ -1205,8 +1205,7 @@ load_identity_file(Identity *id)
 	struct stat st;
 
 	if (stat(id->filename, &st) < 0) {
-		(id->userprovided ? logit : debug3)("no such identity: %s: %s",
-		    id->filename, strerror(errno));
+		debug3("no such identity: %s", id->filename);
 		return NULL;
 	}
 	snprintf(prompt, sizeof prompt,
