diff --git a/arch/arm/boot/dts/amlogic/meson8m2_wetek_core.dtd b/arch/arm/boot/dts/amlogic/meson8m2_wetek_core.dtd
index b4278c2..c704f6f 100644
--- a/arch/arm/boot/dts/amlogic/meson8m2_wetek_core.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8m2_wetek_core.dtd
@@ -303,7 +303,7 @@ void root_func(){
		compatible = "amlogic,amvideocap";
		dev_name = "amvideocap.0";
		status = "okay";
-		reserve-memory = <0x00600000>;//1920*1088*3
+		reserve-memory = <0x01800000>;//3840*2160*3
		reserve-iomap = "true";
	};
 
