diff --git a/init/halt.c b/init/halt.c
index ad12d91..7d498fd 100644
--- a/init/halt.c
+++ b/init/halt.c
@@ -132,7 +132,7 @@ int halt_main(int argc UNUSED_PARAM, char **argv)
 
 	/* Perform action. */
 	rc = 1;
-	if (!(flags & 4)) { /* no -f */
+	if (ENABLE_INIT && !(flags & 4)) { /* no -f */
 //TODO: I tend to think that signalling linuxrc is wrong
 // pity original author didn't comment on it...
 		if (ENABLE_FEATURE_INITRD) {
