diff --git a/setup.py b/setup.py
index d4de292..0a1499a 100644
--- a/setup.py
+++ b/setup.py
@@ -437,16 +437,6 @@ class pil_build_ext(build_ext):
                         _find_library_file(self, "libwebpdemux")):
                     feature.webpmux = "libwebpmux"
 
-        for f in feature:
-            if not getattr(feature, f) and feature.require(f):
-                if f in ('jpeg', 'libz'):
-                    raise ValueError('%s is required unless explicitly disabled'
-                                     ' using --disable-%s, aborting' %
-                                     (f, f))
-                raise ValueError(
-                    '--enable-%s requested but %s not found, aborting.'
-                    % (f, f))
-
         #
         # core library
 
