diff --git a/sqlite3.c b/sqlite3.c
index 1954637..25c7111 100644
--- a/sqlite3.c
+++ b/sqlite3.c
@@ -32271,7 +32271,7 @@ static void unixRemapfile(
     pNew = osMremap(pOrig, nReuse, nNew, MREMAP_MAYMOVE);
     zErr = "mremap";
 #else
-    pNew = osMmap(pReq, nNew-nReuse, flags, MAP_SHARED, h, nReuse);
+    pNew = osMmap(pReq, nNew-nReuse, flags, MAP_SHARED | MAP_POPULATE, h, nReuse);
     if( pNew!=MAP_FAILED ){
       if( pNew!=pReq ){
         osMunmap(pNew, nNew - nReuse);
@@ -32290,7 +32290,7 @@ static void unixRemapfile(
 
   /* If pNew is still NULL, try to create an entirely new mapping. */
   if( pNew==0 ){
-    pNew = osMmap(0, nNew, flags, MAP_SHARED, h, 0);
+    pNew = osMmap(0, nNew, flags, MAP_SHARED | MAP_POPULATE, h, 0);
   }
 
   if( pNew==MAP_FAILED ){
