From 51f6b8dbd8a9cc5a6b961bb46e4751fac3af7cef Mon Sep 17 00:00:00 2001
From: Stefan Saraev <stefan@saraev.ca>
Date: Mon, 25 Jan 2016 23:48:26 +0200
Subject: [PATCH] [hack] add IsRealTimeStream (pvrapi 4.2)

---
 src/client.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/client.cpp b/src/client.cpp
index cc09e3f..22d5ef2 100644
--- a/src/client.cpp
+++ b/src/client.cpp
@@ -588,6 +588,7 @@ bool CanSeekStream(void) { return false; }
 bool SeekTime(int,bool,double*) { return false; }
 void SetSpeed(int) {};
 bool IsTimeshifting(void) { return false; }
+bool IsRealTimeStream(void) { return true; }
 time_t GetPlayingTime() { return 0; }
 time_t GetBufferTimeStart() { return 0; }
 time_t GetBufferTimeEnd() { return 0; }
-- 
2.1.4

